### DISCUSSION TITLE
[HELP] YesSQL Complex Query

### DISCUSSION THREAD

**Discussion by XpandIT-RFID** [OrchardCMS/OrchardCore#9956] [General]
Hello, I am trying to do a query that filters content items of the type "Opportunity"  which uses a custom field and its index.
Until that point I got it working but now I have a string field to search for the Owner name wich is a CustomUserSetting.
The Owner is located in a UserPickerField inside the content type.

So inside the query I need to obtain the CustomUserSettings Name from the userId of the UserPickerField and verify with the filter.SearchBy..

I could query all users then search if there is one but I would have to bring them all to query them after...

Can someone help me with this please? I am more the EntityFramework kind of guy, still learning Dapper and YesSQL.

```C#
            var contentItems = await _orchardSession.Query<ContentItem, ContentItemIndex>()
                .Any(
                    x => x.With<ContentItemIndex>(ci => string.IsNullOrEmpty(filter.SearchBy) || string.Compare(filter.SearchBy, ci.DisplayText, true) == 0))
                    x => x.With<UserPickerFieldIndex>(up => string.IsNullOrEmpty(filter.SearchBy) || [WHAT DO I DO HERE?] ))
                .With<SelectOptionsFieldIndex>(x => string.IsNullOrEmpty(filter.CollaborationArea) || x.ContentField == nameof(OpportunityPart.CollaborationAreas) && x.SelectedValue == filter.CollaborationArea)
                .With<SelectOptionsFieldIndex>(x => string.IsNullOrEmpty(filter.TargetGroup) || x.ContentField == nameof(OpportunityPart.TargetGroups) && x.SelectedValue == filter.TargetGroup)
                .With<SelectOptionsFieldIndex>(x => string.IsNullOrEmpty(filter.Country) || x.ContentField == nameof(OpportunityPart.TargetGroups) && x.SelectedValue == filter.Country)
                .With<SelectOptionsFieldIndex>(x => string.IsNullOrEmpty(filter.District) || x.ContentField == nameof(OpportunityPart.TargetGroups) && x.SelectedValue == filter.District)
                .With<SelectOptionsFieldIndex>(x => string.IsNullOrEmpty(filter.Council) || x.ContentField == nameof(OpportunityPart.TargetGroups) && x.SelectedValue == filter.Council)
                .With<SelectOptionsFieldIndex>(x => string.IsNullOrEmpty(filter.Parish) || x.ContentField == nameof(OpportunityPart.TargetGroups) && x.SelectedValue == filter.Parish)
                .With<UserPickerFieldIndex>(x => string.IsNullOrEmpty(filter.SearchBy))
                .ListAsync();
```

**Comment by XpandIT-RFID** [comment]
![image](https://user-images.githubusercontent.com/83280308/126657411-0805c027-bb50-4915-8a20-cc93c43b98a6.png)
Code picture.

**Comment by XpandIT-RFID** [comment]
@sebastienros help pls 🆘

**Reply by deanmarcussen** [reply]
`= null` should be fine for the string. 

contains you probably want to use something like `x => x.UserId.IsIn(field.UserIds)` assuming you want a `WHERE IN ()` sql clause.

The `.Any()` and `.All()` query methods maybe useful for you as well.

**Reply by XpandIT-RFID** [reply]
I am supposed to search for the username which is one of the CustomUserSettings so I don't have the field.UserIds :/
I didn't want to bring all the users to memory to search for each CustomUserSettings

**Reply by XpandIT-RFID** [reply]
also == null throws this :/
![image](https://user-images.githubusercontent.com/83280308/126684170-ba6299d1-f2b7-45ee-9300-66614d76fc53.png)

**Reply by deanmarcussen** [reply]
what version of YesSql are you using? that looks old.

Regardless if you want to filter on the name property in your CustomUserSettings you will have to make an index for it,

**Reply by XpandIT-RFID** [reply]
I am using Orchard 1.0.0-rc2-16379 which has YesSQL 3.0.7, I created a UserPickerCustomFieldIndex for it which includes the name, so that part I got kind of done... The only problem is that the ContentItem has to be publish for it to get the changes of the username if it ever changes. 
But I don't quite get why I can't check a string for null in the predicate..

**Reply by XpandIT-RFID** [reply]
Because I want it to check the condition only if the filter was filled and this is the only way I know...

**Reply by deanmarcussen** [reply]
oh I think I get what you are asking.
just do this perhaps - you can see this kind of code all through the code base if you look
```
if (!String.IsNullOrEmpty(xxx)
{
   query.WithIndex<Ind>(x => x.xxx);
}

**Reply by XpandIT-RFID** [reply]
Yeah just though of this as well :) 

![image](https://user-images.githubusercontent.com/83280308/126687200-8abe9d9a-d363-4e29-8de1-00c5354ca71a.png)

I like those big queries all going into the database instead of having the ifs but this will do :D 

tks @deanmarcussen 👍

**Reply by XpandIT-RFID** [reply]
@deanmarcussen would it be possible to index CustomUserSettings into the SQL Fields Indexes? Since they use ContentTypes it should be doable right?

**Reply by XpandIT-RFID** [reply]
CustomUserSettings are also ContentItems so I don't quite get why they aren't indexed 🤔

**Reply by deanmarcussen** [reply]
They're a content item stored as json in the user properties `JObject`. totally different thing. 

you can of course write an index for them, and if it is useful for you, you should

but it will be an index based on the `User` document which is not a `ContentItem`

**Reply by XpandIT-RFID** [reply]
Yeah I will be needing to list users so I will probably create a custom user index as well, thanks again @deanmarcussen

**Reply by Skrypt** [reply]
The UserIndex table contains the normalized user name that you are looking for.

**Reply by XpandIT-RFID** [reply]
@Skrypt the name I was looking for is a CustomUserSettings because I didn't want to use it for the login

**Reply by XpandIT-RFID** [reply]
@deanmarcussen after some testing this doesn't quite work... the query need to be inside an all() since the SelectOptionsFieldIndex is only generating one INNER JOIN which then doesn't work... 
But inside the all I cant use the string.IsNullOrEmpty... 

Example:
`SELECT DISTINCT [DefaultTavares_Document].*,[DefaultTavares_Document].[Id] FROM [DefaultTavares_Document]
INNER JOIN [DefaultTavares_ContentItemIndex] AS [ContentItemIndex_a1] ON [ContentItemIndex_a1].[DocumentId] = [DefaultTavares_Document].[Id]
INNER JOIN [DefaultTavares_SelectOptionsFieldIndex] AS [SelectOptionsFieldIndex_a1] ON [SelectOptionsFieldIndex_a1].[DocumentId] = [DefaultTavares_Document].[Id]
WHERE ((([ContentItemIndex_a1].[ContentType] = 'Opportunity') and ([ContentItemIndex_a1].[Published] = 1)) and ([ContentItemIndex_a1].[Latest] = 1))
AND (([SelectOptionsFieldIndex_a1].[ContentField] = 'CollaborationAreas') and ([SelectOptionsFieldIndex_a1].[SelectedValue] = '3ca0f86c-c8c3-4219-a7eb-856932c84fe5'))
AND (([SelectOptionsFieldIndex_a1].[ContentField] = 'TargetGroups') and ([SelectOptionsFieldIndex_a1].[SelectedValue] = '3977bdbe-aeab-4753-85ce-bb25d2be0bcd'))
`

**Comment by RockNHawk** [comment]
You can use another ORM like  NHibernate or EF, then mapping Orchard Table, then Query it :-)

**Comment by XpandIT-RFID** [comment]
Well for those who might see this in the future looking to do something like what I did.
This was the best way I found and the most optimized as well. This way you only ask for the data you want and only get the inner joins that are needed.

Thanks to all that replied and helped!

![image](https://user-images.githubusercontent.com/83280308/127154919-cca7377e-1d32-4729-82bb-60e878620f89.png)
